shm_wrapper/shm_wrapper.c

```diff
@@ -511,159 +546,338 @@ void device_disconnect (int dev_ix)
-	if (dev_ix == -1) {
+	if ((dev_ix = get_dev_ix_from_uid(dev_uid)) == -1) {
 		log_printf(ERROR, "no device at dev_uid = %llu, write failed", dev_uid);
-		return;
+		return -1;
 	}
 	
 	//call the helper to do the actual reading
 	device_write_helper(dev_ix, process, stream, params_to_write, params);
+	return 0;
+}
+
+/*
+ * Send a sub request to dev_handler for a particular device. Takes care of updating the changed bits.
+ * Should only be called by executor and net_handler
+ * Arguments:
+ *    - uint64_t dev_uid: unique 64-bit identifier of the device
+ *    - process_t process: the calling process (will error if not EXECUTOR or NET_HANDLER)
+ *    - uint32_t params_to_sub: bitmap representing params to subscribe to (nonexistent params should have corresponding bits set to 0)
+ * Returns 0 on success, -1 on failure (unrecognized process, or device is not connect in shm)
+ */
+int place_sub_request (uint64_t dev_uid, process_t process, uint32_t params_to_sub)
+{
+	int dev_ix;               //dev_ix that we're operating on
+	uint32_t *curr_sub_map;   //sub map that we're operating on
+	
+	//validate request and obtain dev_ix, sub_map
+	if (process != NET_HANDLER && process != EXECUTOR) {
+		log_printf(ERROR, "calling device_sub_request from process %u", process);
+		return -1;
+	}
+	if ((dev_ix = get_dev_ix_from_uid(dev_uid)) == -1) {
+		log_printf(ERROR, "no device at dev_uid = %llu, sub request failed", dev_uid);
+		return -1;
+	}
+	curr_sub_map = (process == NET_HANDLER) ? dev_shm_ptr->net_sub_map : dev_shm_ptr->exec_sub_map;
+	
+	//wait on sub_map_sem
+	my_sem_wait(sub_map_sem, "sub_map_sem");
+	
+	//only fill in params_to_sub if it's different from what's already there
+	if (curr_sub_map[dev_ix + 1] != params_to_sub) {
+		curr_sub_map[dev_ix + 1] = params_to_sub;
+		curr_sub_map[0] |= (1 << dev_ix); //turn on corresponding bit
+	}
+	
+	//release sub_map_sem
+	my_sem_post(sub_map_sem, "sub_map_sem");
+	
+	return 0;
+}
+
+/*
+ * Get current subscription requests for all devices. Should only be called by dev_handler
+ * Arguments:
+ *    - uint32_t *sub_map: bitwise OR of the executor and net_handler sub_maps that will be put into this provided buffer
+ * No return value.
+ */
+void get_sub_requests (uint32_t *sub_map)
+{
+	//wait on sub_map_sem
+	my_sem_wait(sub_map_sem, "sub_map_sem");
+	
+	//bitwise or the changes into sub_map[0]; if no changes then return
+	sub_map[0] = dev_shm_ptr->net_sub_map[0] | dev_shm_ptr->exec_sub_map[0];
+	if (sub_map[0] == 0) {
+		my_sem_post(sub_map_sem, "sub_map_sem");
+		return;
+	}
+	
+	//bitwise OR each valid element together into sub_map[i]
+	for (int i = 1; i < MAX_DEVICES + 1; i++) {
+		if (dev_shm_ptr->catalog & (1 << (i - 1))) { //if device exists
+			sub_map[i] = dev_shm_ptr->net_sub_map[i] | dev_shm_ptr->exec_sub_map[i];
+		} else {
+			sub_map[i] = 0;
+		}
+	}
+	
+	//reset the change indicator eleemnts in net_sub_map and exec_sub_map
+	dev_shm_ptr->net_sub_map[0] = dev_shm_ptr->exec_sub_map[0] = 0;
```

Comment:
I like the design of these 2 functions `place_sub_request` and `get_sub_request` because executor and dev_handler can call them both non stop